demo/sources/CarouselComponent.swift

```diff
@@ -0,0 +1,170 @@
+ *  distributed with this work for additional information
+ *  regarding copyright ownership.  The ASF licenses this file
+ *  to you under the Apache License, Version 2.0 (the
+ *  "License"); you may not use this file except in compliance
+ *  with the License.  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing,
+ *  software distributed under the License is distributed on an
+ *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  KIND, either express or implied.  See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ */
+
+import Foundation
+import HubFramework
+
+/**
+ *  A component that renders an array of child components as a horizontally scrollable carousel
+ *
+ *  The carousel uses the `children` property of a `HUBComponentModel` to determine what items
+ *  it should contain. It makes the assumption that the children are all rendered using the same
+ *  component, to simplify its layout calculations.
+ */
+class CarouselComponent: NSObject, HUBComponentWithChildren, HUBComponentWithRestorableUIState, HUBComponentViewObserver, UICollectionViewDataSource, UICollectionViewDelegateFlowLayout {
+    var view: UIView?
+    var childDelegate: HUBComponentChildDelegate?
+    
+    private lazy var collectionView = UICollectionView(frame: .zero, collectionViewLayout: CarouselComponent.makeCollectionViewLayout())
+    private var model: HUBComponentModel?
+    private var itemSize: CGSize?
+    private var cellReuseIdentifier: String { return "cell" }
+    
+    // MARK: - HUBComponent
+
+    var layoutTraits: Set<HUBComponentLayoutTrait> {
+        return [.fullWidth]
+    }
+
+    func loadView() {
+        collectionView.register(HUBComponentCollectionViewCell.self, forCellWithReuseIdentifier: cellReuseIdentifier)
+        collectionView.dataSource = self
+        collectionView.delegate = self
+        collectionView.backgroundColor = .white
+        collectionView.showsHorizontalScrollIndicator = false
+        collectionView.showsVerticalScrollIndicator = false
+        
+        view = collectionView
+    }
+
+    func preferredViewSize(forDisplaying model: HUBComponentModel, containerViewSize: CGSize) -> CGSize {
+        let itemSize = calculateItemSize(forModel: model, containerViewSize: containerViewSize)
+        return CGSize(width: containerViewSize.width, height: itemSize.height)
+    }
+
+    func prepareViewForReuse() {
+        itemSize = nil
+    }
+
+    func configureView(with model: HUBComponentModel, containerViewSize: CGSize) {
+        self.model = model
+        itemSize = calculateItemSize(forModel: model, containerViewSize: containerViewSize)
+        collectionView.reloadData()
+    }
+    
+    // MARK: - HUBComponentWithRestorableUIState
+    
+    func currentUIState() -> Any? {
+        return collectionView.contentOffset
```

Comment:
I'm a bit rusty on the swift/obj-c interop – in obj-c structs these structs would need to be boxed (as the method takes an `id`). Will this work?